server/server.go

```diff
@@ -0,0 +1,100 @@
+package server
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/vault/api"
+)
+
+// New returns an unstarted HTTP server with health and proxy handlers.
+func New(logger *log.Logger, vaultClient *api.Client) *http.Server {
+	mux := http.ServeMux{}
+	mux.HandleFunc("/_health", healthHandler)
+	mux.HandleFunc("/", proxyHandler(logger, vaultClient))
+	srv := http.Server{
+		Addr:    "127.0.0.1:8200",
+		Handler: &mux,
+	}
+
+	return &srv
+}
+
+// The proxyHandler is based on the Send function from Vault Agent's proxy:
+// https://github.com/hashicorp/vault/blob/22b486b651b8956d32fb24e77cef4050df7094b6/command/agent/cache/api_proxy.go
+func proxyHandler(logger *log.Logger, vaultClient *api.Client) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		switch r.Method {
+		case "GET":
+			cl, err := vaultClient.Clone()
+			if err != nil {
+				http.Error(w, err.Error(), 500)
+				return
+			}
+			cl.SetToken(vaultClient.Token())
+
+			body, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				http.Error(w, err.Error(), 500)
+				return
+			}
+
+			logger.Printf("Proxying %s %s\n", r.Method, r.URL.Path)
+			resp, err := proxyRequest(cl, r, body)
+			if err != nil {
+				http.Error(w, err.Error(), 502)
+				return
+			}
+
+			err = resp.Write(w)
+			if err != nil {
+				http.Error(w, "failed to write response back to requester", 500)
+				return
+			}
+			logger.Printf("Successfully proxied %s %s\n", r.Method, r.URL.Path)
+		default:
+			http.Error(w, fmt.Sprintf("unsupported method for vault-lambda-extension: %s", r.Method), 400)
+			return
+		}
+	}
+}
+
+func proxyRequest(client *api.Client, r *http.Request, body []byte) (*api.Response, error) {
+	// http.Transport will transparently request gzip and decompress the response, but only if
+	// the client doesn't manually set the header. Removing any Accept-Encoding header allows the
+	// transparent compression to occur.
+	r.Header.Del("Accept-Encoding")
+	client.SetHeaders(r.Header)
```

Comment:
I don't think you need to call SetHeaders here, though if you remove this we should sort out whether X-Vault-Request should be populated by your proxy.  Actually we need to work that out regardless.

Reply:
One use case I had in mind was setting namespace in the request headers, and I assumed there would be others I didn't know about so thought it would be best to just honour the original HTTP request as closely as possible. Do you think we should handle cases like namespace explicitly?

Good question about the `X-Vault-Request` header, I need to learn a bit more about the design/intention of that before I can answer your question.

Reply:
Oh I agree that we should honour the incoming request's headers, I'm just saying you don't need to call client.SetHeaders for that to happen.  You can simply set fwReq.Headers = r.Header.  I don't know about namespaces, but if you don't have a specific requirement regarding them I probably wouldn't pay them any mind, just pass the requests through largely unchanged and everything should be fine.

Reply:
Ahh that makes sense, yeah that's better, thanks.